feat: add resource property to token request payload

Szik opened the pull request:
## Description

add property `resource` to oauth token request
this PR adds/changes the following:
- `CamundaClientEnvironmentVariables.java` - Added `CAMUNDA_TOKEN_RESOURCE`
-  `ZeebeClientEnvironmentVariables.java` - Added `ZEEBE_TOKEN_RESOURCE`
-  `OAuthCredentialsProvider.java` (both clients) - Added resource to payload
-  `OAuthCredentialsProviderBuilder.java` (both clients) - Added resource methods
-  Test files - Updated for testing new functionality

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [x] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #
add resource property to token request payload to zeebe/camunda client #34964

Szik committed:
feat: fix commit error

ThorbenLindhauer reviewed (changes requested):
Overall it's a good code contribution. I like that you are following the established structures, also the test coding convention with given-when-then.

Besides the detailed remarks, two more points:

* Currently, we only have test coverage for the old `ZeebeClient` (note that your test changes are all in the `io.camunda.zeebe` package. Please also test the `CamundaClient` changes (we have the same test classes there)
* Please also consider writing docs for this, e.g. here for the optional authentication configuration: https://docs.camunda.io/docs/next/apis-tools/java-client/authentication/
  * I'd also be okay that we write this later holistically for all places that we are adding the parameter to; but let's not forget this; config options without docs don't exist in a way.

ThorbenLindhauer commented:
Adding one more point to the review:

* Please also expose the parameter in the Spring SDK (e.g. you can check this PR where you need to make changes: https://github.com/camunda/camunda/pull/34220)

Szik committed:
feat: aouth resource parameter
- add resource property to CamundaClientAuthProperties
- updated CredentialsProviderConfiguration to pass resource to oauth
- add environment variable mapping via zeebe.token.resource + tests
- fixed wiremock tests (resource not part of response)
- removed redundant test SslClientCertAuthTest
- enable spring SDK to use oauth resource parameter via: camunda.client.auth.resource
- properties: camunda.client.auth.resource=value
- environment variaböle: ZEEBE_TOKEN_RESOURCE=value

ThorbenLindhauer commented on clients/java/src/test/java/io/camunda/client/OAuthCredentialsProviderTest.java:
```diff
@@ -495,6 +566,45 @@ private void mockTokenRequest(final boolean withAssertion) {
     }
   }
 
+  private void mockTokenRequestWithResource(final boolean withAssertion, final String resource) {
```
Could you please let `mockTokenRequest(final boolean withAssertion)` call this method here, similar to how you did it with the tests in the zeebe package? That reduces code duplication.

ThorbenLindhauer commented on clients/java/src/test/java/io/camunda/client/OAuthCredentialsProviderTest.java:
```diff
@@ -460,6 +483,54 @@ private void mockCredentials(final String token, final String scope) {
     }
   }
 
+  private void mockCredentialsWithResource(
```
Could you please let `mockCredentials(final String token, final String scope)` call this method here, similar to how you did it with the tests in the zeebe package? That reduces code duplication.

ThorbenLindhauer reviewed (changes requested):
Thanks for making the changes. I have some more thoughts around the tests.

Szik commented on clients/spring-boot-starter-camunda-sdk/src/test/java/io/camunda/spring/client/config/CredentialsProviderResourceEnvTest.java:
```diff
@@ -0,0 +1,52 @@
+    })
+@TestPropertySource(properties = {"zeebe.token.resource=https://api.example.com"})
+@EnableConfigurationProperties(CamundaClientProperties.class)
+public class CredentialsProviderResourceEnvTest {
```
yes, that was on purpose to verify backward compatibility.
if you think its not needed here, i can remove it but if you think its okay to have it, i will extend the test coverage.

Szik committed:
feat: add resource property to token request payload
- refactored OAuthCredentialsProviderTest to make mockTokenRequest()
  and mockCredentials() delegate to mockCredentialsWithResource()
  with null resource
- Added TokenResource test to CamundaClientPropertiesPostProcessorTest
- Updated token-resource.yaml to use legacy Zeebe properties

Szik committed:
fix: removed/resolved dublicated resource properties for tests

ThorbenLindhauer reviewed (changes requested):
Sorry, I think you are resolving this in the wrong direction and it leaves the test cases more chaotic if we keep it like that. Please let me know if I am overlooking something.

Szik committed:
fix: rm env test, added test to post processor

ThorbenLindhauer reviewed (approved):
Thanks for making all the changes, I think we are good now :)

backport-action commented:
Created backport PR for `stable/8.6`:
- #35227 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-34963-to-stable/8.6
git worktree add --checkout .worktree/backport-34963-to-stable/8.6 backport-34963-to-stable/8.6
cd .worktree/backport-34963-to-stable/8.6
git reset --hard HEAD^
git cherry-pick -x 6623917205c08ec4562595a8cd7660385bd1ea21 930945bd6b19d3904c743a9c80de4acd60707027 4ef45b18ce15b8691951f40d4ee582ac9d52c688 3a6101e5e7804eaf3395af0cea68495e98a0c30c 9a07dc2b9c7b3d6d20a66ff948b13b0d2d64c4da
git push --force-with-lease
```
